Add contributing guide and development docs (#502)
celeritas-project/celeritas: doc/appendices/development.rst

Development Guide
*****************

Guiding principles
==================

Maximize encapsulation
----------------------

Encapsulation is about making a piece of code into a black box. The fewer lines
connecting these black boxes, the more maintainable the code. Black boxes can
often be improved internally by making tiny black boxes inside the larger black
box.

Motivation:

- Developers don't have to understand implementation details when looking at a
  class interface.
- Compilers can optimize better when dealing with more localized components.
- Good encapsulation allows components to be interchanged easily because they
  have well-defined interfaces.
- Pausing to think about how to minimize input and output from an algorithm can
  improve make it easier to write.

Applications:

> - Refactor large functions (> 50 statements ish?) into small functors that take

Reviewer: ```suggestion
- Refactor large functions (> 50-ish statements?) into small functors that take
```

  "invariant" values (the larger context) for constructors and use
  ``operator()`` to transform some input into the desired output
- Use only ``const`` data when sharing. Non-const shared data is almost like
  using global variables.
- Use ``OpaqueId`` instead of integers and magic sentinel values for
  integer identifiers that aren't supposed to be arithmetical.

Examples:

- Random number sampling: write a unit sphere sampling functor instead of

>   replicating a polar-to-cartesian transform in a thousand places

Reviewer: ```suggestion
  replicating a polar-to-cartesian transform in a thousand places.
```

- Cell IDs: Opaque IDs add type safety so that you can't accidentally convert a

>   cell identifier into a double or switch a cell and material ID. Also makes

Reviewer: ```suggestion
  cell identifier into a double or switch a cell and material ID. It also makes
```

  code more readable of course.

Maximize code reuse
-------------------

No explanation needed.

Minimize compile time
---------------------

> Code performance is important, but so is developer time. When possible,

Reviewer: ```suggestion
Code performance is important but so is developer time. When possible,
```

minimize the amount of code touched by NVCC. (NVCC's error output is also
rudimentary compared to modern clang/gcc, so that's another reason to prefer
them compiling your code.)

Prefer single-state classes
---------------------------

As much as possible, make classes "complete" and valid after calling the
constructor. Don't have a series of functions that have to be called in a

> specific order to put the class in a workable state.

Reviewer 2: ```suggestion
specific order to put the class in a workable state.  (And when it is not possible,  code must be put in place to automatically detect (and warn the developer) if the specific order is not respected).
```

When a class has a single function (especially if you name that function
``operator()``), its usage is obvious. The reader also doesn't have to know
whether a class uses ``doIt`` or ``do_it`` or ``build``.

When you have a class that needs a lot of data to start in a valid state, use a
``struct`` of intuitive objects to pass the data to the class's constructor.
The constructor can do any necessary validation on the input data.

Learn from the pros
-------------------

Other entities devoted to sustainable programming have their own guidelines.
The `ISO C++ guidelines`_ are very long but offer a number of insightful
suggestions about C++ programming. The `Google style guide`_ is a little more
targeted toward legacy code and large production environments, but it still
offers good suggestions.

.. _ISO C++ guidelines: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines
.. _Google style guide: https://google.github.io/styleguide/cppguide.html

Test thoroughly
---------------

Functions should use programmatic assertions whenever assumptions are made:

- Use the ``CELER_EXPECT(x)`` assertion macro to test preconditions about
  incoming data or initial internal states

> - Use ``CELER_ASSERT(x)`` to express an assumption internal to a function (e.g.

Reviewer: ```suggestion
- Use ``CELER_ASSERT(x)`` to express an assumption internal to a function (e.g.,
```

the `Draco project style`_ of which `Tom Evans`_ was primary author and which
became the style standard for the GPU-enabled Monte Carlo code `Shift`_.

.. _Draco project style: https://github.com/lanl/Draco/wiki/Style-Guide
.. _Tom Evans: https://github.com/tmdelellis
.. _Shift: https://doi.org/10.1016/j.anucene.2019.01.012

Formatting
----------

Formatting is determined by the clang-format file inside the top-level
directory. One key restriction is the 80-column limit, which enables multiple
code windows to be open side-by-side. Generally, statements longer than 80
columns should be broken into sub-expressions for improved readability anyway
-- the ``auto`` keyword can help a lot with this.

There's a certain amount of decorations (separators, Doxygen comment structure,
etc.) that is standard throughout the code. Use the ``celeritas-gen.py`` script
(in the ``scripts/dev`` directory) to generate skeletons for new files, and use
existing source code as a guide to how to structure the decorations.

Symbol names
------------

Functions should be verbs; classes should be names. As in standard Python
(PEP-8-compliant) code, classes should use ``CapWordsStyle`` and variables use

> ``snake_case_style``.

Reviewer 2: You may want to add a mention of the convention of members of structs and classes.

  file. These should only be in the main ``/test`` directory.

If there are only a few short inline methods (and especially if it's for a small
class) they can be included at the bottom of the main header file under a
suitable
separator that demarcates the declarations from the definitions.

Device compilation
------------------

All ``__device__`` and ``__global__`` code must be compiled with NVCC or
HIPCC to generate device objects. However, code that merely uses CUDA API calls
such as
``cudaMalloc`` does *not* have to be compiled with NVCC. Instead, it only has to
be linked against the CUDA runtime library and include ``cuda_runtime_api.h``.
The platform-agnostic Celeritas include file to use is
``corecel/device_runtime_api.h``.
Note that VecGeom compiles differently when run
through NVCC (macro magic puts much of the code in a different namespace) so
its inclusion must be very carefully managed.

Since NVCC is slower and other compilers' warning/error output is more
readable, it's preferable to use NVCC for as little compilation as possible.
Furthermore, not requiring NVCC lets us play nicer with downstream libraries

> and front-end apps. Host code will not be restricted to the minimum version
> supported by NVCC.

Reviewer 2: Do you mean 
```suggestion
and front-end apps. Host code will not be restricted to the maximum C++ standard version
supported by NVCC.
```
?

  use shared pointers instead to make the ownership semantics clear. When
  interfacing with older libraries such as Geant4, try to use ``unique_ptr``
  and its ``release``/``get`` semantics to indicate the transfer of pointer
  ownership.
- Since we don't yet support C++17's ``string_view`` it's OK to use ``const
  char*`` to indicate a read-only string.

Memory is always managed from host code, since on-device data management can be
tricky, proprietary, and inefficient. There are no shared or unique pointers,
but there is less of a need because memory management semantics are clearer.
Device code has exceptions from the rules above:

- In low-level device-compatible code (such as a ``TrackView``), it is OK to
  return a pointer from a function to indicate that the result may be undefined
  (i.e., the pointer is null) or a non-owning **reference** to valid memory.
  This is used in the ``StackAllocator`` to indicate a failure to allocate new
  memory, and in some accessors where the result is optional.
- The rule of passing references to complex data does not apply to CUDA
  ``__global__`` kernels, because device code cannot accept references to host

>   memory. Instead, kernels parameters should copy by value or provide raw
>   pointers to device memory. Indicate that the argument should not be used
>   inside the kernel can prefix it with ``const``, so the CUDA compiler places

Reviewer 3: ```suggestion
  memory. Instead, kernel parameters should copy by value or provide raw
  pointers to device memory. Indicate that the argument should not be used
  inside the kernel can prefix it with ``const``, so the CUDA compiler places
```
and fix wording

  ``TrackFooState``.

Data
  A standalone, plain-old-data struct to data owned by another class (e.g., a
  Params class) but stored on the GPU. This struct is used to transfer GPU
  pointers and other kernel parameters between host and device. A Data
  struct can hold other Data structs as data members. Inside unit tests for
  debugging, Data can reference *host* data if the corresponding functions
  being called are also on-host. Defining Data structs in separate files
  from the memory management classes means that NVCC doesn't have to include
  those headers, speeding up compilation and perhaps allowing the host code to
  use newer implementations of the C++ standard.

TrackView
  Device-instantiated class that provides read/write access to the data for a
  single track, in the spirit of `std::string_view` which adds functionality to
  data owned by someone else. It combines the state variables and model
  parameters into a single class. The constructor always takes const references
  to ParamsData and StatePointers as well as the track ID. It encapsulates
  the storage/layout of the state and parameters, as well as what (if any) data
  is cached in the state.

View

>   Device-instantiated class with read/write access for data shared across

Reviewer 2: Is the 'Device' limitation still in place?  I.e. is code using Views not able to be tested on CPU?

Author: Nope this was just a missed update from the year+-old document :) fixed
